File: internal/console/ui/app/screens/today.js
/* ==========================================================================
   What needs a person today.

   # THE SCREEN THAT ANSWERS "WHERE DO I LOOK"

   Fourteen screens each answer one question well, and none of them answered
   that one — so opening this console cold meant already knowing which screen
   held the thing that was wrong. The screens somebody opens are the ones they
   already suspect, and the defect nobody suspects is the one that stays.

   # IT IS EMPTY WHEN NOTHING IS WRONG, AND THAT IS THE FEATURE

   A wall of figures always has something to look at, which is how a screen
   stops being read. This one is a list of things that are TRUE and want
   somebody, so a quiet platform draws a sentence and nothing else — and a line
   appearing on it means something.

   THE EMPTY STATE IS THE SERVER'S SENTENCE, because "there is nothing to do"
   and "this failed to load" look identical, and this is the one screen where
   the first would be read as good news.

   # EVERY LINE IS ONE CLICK FROM THE SCREEN THAT EXPLAINS IT

   Nothing here carries a threshold, a chart or an argument: those live on the
   screen that owns the finding, with the numbers that produced it. This says
   what is true and where to go, which is all a person needs to decide whether
   tonight is the night.
   ========================================================================== */

import { esc } from '../dom.js';
import { get } from '../request.js';
import { txt } from '../../assets/language.js';
import { goTo } from '../routes.js';

/* THE RAIL'S OWN NAME FOR A SCREEN, so a finding says "Reported content" rather
   than the route `reports`. It is imported from the file that imports this one,
   which is a cycle — and a safe one, because `sectionById` is only ever CALLED
   while a finding is being drawn, long after both modules have finished
   evaluating. Reading it at module level would be the version that breaks. */
import { sectionById } from '../sections.js';

/* WHAT EACH FINDING SAYS, in one sentence for one of them and another for
   several. A count of one followed by a plural is the defect this console has
   shipped once — "1 trilhas" — and a sentence is what an operator reads
   anyway: `3` beside a label is a number to decode, and "three questions the
   analysis condemned are still being asked" is already the thought.

   EVERY SENTENCE IS ONE LITERAL ON ONE LINE, however long. `language_test.go`
   reads this map — the strings are looked up through a variable, so
   `check-interface` cannot see them — and it takes the first literal of a
   concatenation, which would ask the dictionary for half a sentence.

   THE KEYS ARE THE SERVER'S WORDS. A kind this list does not know is drawn as
   itself rather than folded into the nearest — the closed-list rule every other
   screen here follows, so a finding added on the server appears as something
   nobody has styled rather than as the wrong sentence. */
const SAYS = {
  'questions-still-asked': {
    one: 'A question the analysis condemned is still being asked.',
    many: '%d questions the analysis condemned are still being asked.',
  },
  'reports-waiting': {
    one: 'A student wrote in about a lesson and nobody has answered.',
    many: '%d students wrote in about a lesson and nobody has answered.',
  },
  'roles-without-a-second-factor': {
    one: 'Somebody holds a role and has no second factor, so it opens nothing.',
    many: '%d people hold a role and have no second factor, so it opens nothing.',
  },
  'roles-never-used': {
    one: 'A role was granted and has never been used.',
    many: '%d roles were granted and have never been used.',
  },
  'jobs-that-failed': {
    one: 'A scheduled job failed the last time it ran.',
    many: '%d scheduled jobs failed the last time they ran.',
  },
  'jobs-adrift': {
    one: 'A job started and never said it finished.',
    many: '%d jobs started and never said they finished.',
  },
  'jobs-that-never-ran': {
    one: 'A job that can be started has never recorded a run.',
    many: '%d jobs that can be started have never recorded a run.',
  },
  'could-not-be-checked': {
    one: 'One of the things this screen checks could not be read, so its line is missing rather than empty.',
    many: '%d of the things this screen checks could not be read, so their lines are missing rather than empty.',
  },
};

export default async function today(section) {
  const el = document.createElement('div');
  el.className = 'view';

  el.innerHTML =
    '<header class="view-head">' +
      '<span class="eyebrow mono">' + esc(txt('Attend')) + '</span>' +
      '<h1>' + esc(txt('What needs you')) + '</h1>' +
      '<p>' + esc(txt('Everything on this platform that is true right now and wants a person. It is empty on the days nothing is — which is most of them, and is the reason a line appearing here means something.')) + '</p>' +
    '</header>' +
    '<div id="body" aria-live="polite"><p class="checking">' + esc(txt('Reading…')) + '</p></div>';

  const body = el.querySelector('#body');

  let answer;
  try {
    answer = await get('/console/api/v1/today');
  } catch (e) {
    body.innerHTML = '<section class="block"><p class="none">' + esc(txt(e.message)) + '</p></section>';
    return { title: section.name, el };
  }

  const found = answer.findings || [];

  body.innerHTML =
    (found.length === 0
      ? '<section class="block"><p class="none">' +
        esc(txt(answer.nothing_to_do || '')) + '</p></section>'
      : '<section class="block"><ul class="findings">' +
          found.map(row).join('') +
        '</ul></section>') +

    /* SAID WHETHER OR NOT ANYTHING IS WRONG. Somebody looking at a console home
       reasonably expects it to tell them when something breaks, and a screen
       that let the expectation stand is one they would rely on at the worst
       possible moment (K-08). It is the server's sentence. */
    '<section class="block">' +
      '<div class="block-top"><h2>' + esc(txt('This is not an alarm')) + '</h2></div>' +
      '<p class="aside">' + esc(txt(answer.not_an_alarm || '')) + '</p>' +
    '</section>';

  /* THE WHOLE ROW IS THE CONTROL. A finding is a thing to act on, so the target
     is the sentence rather than a "view" link beside it — and it is a `button`
     rather than an anchor because the router is the thing that moves, which is
     what every other in-console navigation does. */
  body.querySelectorAll('[data-where]').forEach((one) => {
    one.addEventListener('click', () => goTo('/' + one.dataset.where));
  });

  return { title: section.name, el };
}

function row(finding) {
  const says = SAYS[finding.kind];

  /* A KIND NOBODY HAS WRITTEN A SENTENCE FOR IS STILL SHOWN. It is a real
     finding that a screen simply does not have words for yet, and hiding it
     would make this screen quietly incomplete — which is the one thing it may
     not be. The count and the server's own word are what there is. */
  const said = says
    ? (finding.count === 1 ? txt(says.one) : txt(says.many).replace('%d', finding.count))
    : finding.kind + ' — ' + finding.count;

  /* A FINDING WITH NOWHERE TO GO IS NOT A BUTTON. `could-not-be-checked` has no
     screen of its own: it is about this one. A control that navigated nowhere
     would be worse than a sentence. */
  if (!finding.where) {
    return '<li class="finding finding-quiet"><span>' + esc(said) + '</span></li>';
  }

  return '<li class="finding">' +
    '<button type="button" class="finding-go" data-where="' + esc(finding.where) + '">' +
      '<span class="finding-said">' + esc(said) + '</span>' +
      '<span class="finding-where mono">' + esc(nameOf(finding.where)) + '</span>' +
    '</button>' +
  '</li>';
}

// What the rail calls the screen a finding points at, falling back to the route
// where a section has been renamed out from under a finding the server still
// sends — visible rather than blank.
function nameOf(where) {
  const section = sectionById(where);
  return section ? txt(section.name) : where;
}

File: internal/console/today.go
package console

import (
	"context"
	"fmt"
	"net/http"

	"github.com/codeschool-ing/schooling/internal/platform/web"
)

/* ==========================================================================
   What needs a person today.

   # THE QUESTION FOURTEEN SCREENS COULD NOT ANSWER

   Every screen here answers one question well and says its scope. None of them
   answers "what should I look at?", so opening this console cold meant already
   knowing which screen held the thing that was wrong — and the screens somebody
   opens are the ones they already suspect. The defect nobody suspects is the
   one that stays.

   # IT IS FINDINGS AND NOT FIGURES, WHICH IS THE WHOLE DESIGN

   A dashboard of numbers is a screen with no question, and this console's rule
   is that a screen says what it is about (K-18). "1,204 students" under no
   scope, no population and no date is exactly what that decision exists against
   — and there is nowhere on a tile to put the sentence the cohorts screen needs
   a paragraph for.

   So every line here is something that is TRUE and wants a person: a question
   still being asked that the analysis condemned, a night that did not run, a
   report nobody answered. On a platform where nothing is wrong this screen is
   empty, which is the right amount of information and the opposite of a wall of
   figures that always has something to look at.

   # EVERY FINDING IS A FACT AND ALMOST NONE OF THEM IS A THRESHOLD

   `granted and never used` is a fact. `no second factor` is a fact. `waiting to
   be answered` is a fact. Only the adrift run is a judgement, and its threshold
   is the one the jobs screen already draws and the store already decided —
   K-16, one layer along: a number that decided something appears beside the
   thing it decided, and never in two places with a chance to disagree.

   # AND IT ASKS THE SAME FUNCTIONS THE SCREENS ASK

   `Operators` and `Jobs` are the seams the staff and jobs screens are built on,
   passed in here unchanged. A home screen with its own idea of "has this person
   ever opened the console" would be a second definition of the same fact, and
   the two would disagree on exactly the row somebody is trying to understand.
   That is the mistake `personOf` is shared to avoid, one report over.

   # IT IS NOT AN ALARM AND SAYS SO

   K-08: alerts have to reach a phone when this console is down, which is when
   they matter. This is a screen somebody opens, and everything on it is true
   for as long as nobody acts on it — which is a different job from waking
   anybody up.
   ========================================================================== */

// Finding is one thing that is true and wants somebody.
type Finding struct {
	// Kind is what it is, in this console's own words, and it is what the
	// screen keys its wording off. A kind this interface does not know is drawn
	// as itself rather than folded into another, the way every closed list here
	// behaves.
	Kind string

	// Count is how many, and it is never zero: a finding that is not true is
	// not sent. "Nothing is wrong" is an empty list rather than a column of
	// zeroes, because a zero is a thing to read and this screen is for the
	// things that are not.
	Count int

	// Where is the section id the reader should go to, so every finding is one
	// click from the screen that owns it and can explain it.
	Where string
}

/*
Today is what the home screen reads.

	THE TWO COUNTS ARE THEIR OWN SEAMS AND THE OTHER TWO ARE NOT. `Condemned`
	and `Waiting` are questions no existing screen asks — they are across every
	school, where the screens are about one — so they arrive as functions this
	package names. `Operators` and `Jobs` already exist and are handed over as
	they are, because asking the same function is what stops two screens
	disagreeing about one fact.
*/
type Today struct {
	// Condemned is how many questions the analysis called inverted that are
	// still being given to students, across every school.
	Condemned func(ctx context.Context) (int, error)

	// Waiting is how many reports nobody has answered, across every school.
	Waiting func(ctx context.Context) (int, error)

	// Operators is the staff screen's own reader, unchanged.
	Operators Operators

	// Jobs is the jobs screen's own, unchanged.
	Jobs Jobs
}

type findingBody struct {
	Kind  string `json:"kind"`
	Count int    `json:"count"`

	// Empty for a finding about this screen rather than another, which the
	// interface draws as a sentence instead of a control.
	Where string `json:"where,omitempty"`
}

// TodayHandler answers the home screen. It reads and never writes, so it
// carries no audit seam and no second rank — every staff role may look, which
// is the same rule the aggregates are under.
type TodayHandler struct{ today Today }

func NewTodayHandler(today Today) *TodayHandler { return &TodayHandler{today: today} }

func (h *TodayHandler) Routes(mux *http.ServeMux) {
	mux.HandleFunc("GET /console/api/v1/today", h.findings)
}

/*
findings gathers what is true right now.

	A READER THAT FAILS IS A FINDING OF ITS OWN AND NOT A DEAD SCREEN. Four
	questions are asked and any of them can fail; refusing the whole page
	because the jobs table was unreachable would hide the three findings that
	did arrive — on the one screen whose job is to say what is wrong. So a
	failure is carried as `unreadable`, which the screen draws as the finding it
	is: something here could not be checked, and the number beside it is not
	zero because nobody looked.
*/
func (h *TodayHandler) findings(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	log := web.LoggerFrom(ctx)

	var out []Finding
	var unreadable int

	add := func(kind, where string, n int) {
		if n > 0 {
			out = append(out, Finding{Kind: kind, Count: n, Where: where})
		}
	}

	ask := func(what string, of func(ctx context.Context) (int, error)) int {
		if of == nil {
			return 0
		}
		n, err := of(ctx)
		if err != nil {
			log.Error("reading what needs a person", "finding", what, "error", err)
			unreadable++
			return 0
		}
		return n
	}

	add("questions-still-asked", "questions", ask("questions", h.today.Condemned))
	add("reports-waiting", "reports", ask("reports", h.today.Waiting))

	people, err := h.operators(ctx)
	if err != nil {
		log.Error("reading what needs a person", "finding", "staff", "error", err)
		unreadable++
	} else {
		add("roles-without-a-second-factor", "staff", people.noFactor)
		add("roles-never-used", "staff", people.neverUsed)
	}

	jobs, err := h.jobs(ctx)
	if err != nil {
		log.Error("reading what needs a person", "finding", "jobs", "error", err)
		unreadable++
	} else {
		add("jobs-that-failed", "jobs", jobs.failed)
		add("jobs-adrift", "jobs", jobs.adrift)
		add("jobs-that-never-ran", "jobs", jobs.never)
	}

	add("could-not-be-checked", "", unreadable)

	/* THE JSON NAMES ARE A TYPE OF THEIR OWN, which is this package's shape for
	   every answer — `stepBody`, `cohortBody` — and the reason is what happened
	   without it: `Finding` went out as `Kind`, `Count` and `Where`, the screen
	   read `kind` and `count`, and every row drew "undefined — undefined". It
	   rendered perfectly and said nothing, which is the failure this whole
	   console is written against. Found by opening the screen.

	   THE CONVERSION IS SAFE BECAUSE THE COMPILER CHECKS IT, which is the
	   reason `stepBody` is written the same way one file over: the two types
	   having the same fields is what makes this one expression rather than
	   three, and the day they stop matching this line stops building rather
	   than quietly dropping a field. */
	body := make([]findingBody, 0, len(out))
	for _, one := range out {
		body = append(body, findingBody(one))
	}

	web.JSON(w, http.StatusOK, map[string]any{
		// EMPTY RATHER THAN NIL, so a quiet platform is an empty list and not a
		// field that failed to arrive — the distinction the funnel's `Measured`
		// makes, one screen along.
		"findings": body,

		/* AND WHAT THIS SCREEN IS NOT, said by the thing that knows. Somebody
		   looking at a console home reasonably expects it to tell them when
		   something breaks; K-08 says that cannot be this, and a screen which
		   let the expectation stand would be one people rely on at exactly the
		   wrong moment. */
		"not_an_alarm": "Everything here is true for as long as nobody acts on it. It is " +
			"not an alarm and cannot be: an alert has to reach a phone when this console " +
			"is down, which is exactly when it matters.",

		// WHY IT IS EMPTY, WHEN IT IS. An empty screen and a screen that failed
		// to load look identical, and this is the one place that would be read
		// as good news.
		"nothing_to_do": "Nothing on this platform is asking for a person right now. This " +
			"screen is empty when that is true, which is most days — it is the screens " +
			"under Measure that are worth opening when nothing is wrong.",
	})
}

type roster struct{ noFactor, neverUsed int }

/*
operators counts the two findings a roster carries.

	NEITHER IS A THRESHOLD. A role with no second factor opens nothing, which is
	a fact about the door. A role granted and never used is access nobody is
	missing — also a fact, and the row an access review exists to find.

	THE STAFF SCREEN'S NINETY DAYS IS DELIBERATELY NOT HERE. That number is a
	reading aid on a screen full of dates; here it would decide whether a line
	appears at all, which makes it a threshold that judges (K-16) rather than
	one that helps somebody read. `never` needs no number and is the stronger
	finding anyway.

	AND A REVOKED ROW IS NOT A FINDING. Somebody who left has no access to
	review; counting them would put a permanent number on this screen that no
	action can ever clear, which teaches people to ignore it.
*/
func (h *TodayHandler) operators(ctx context.Context) (roster, error) {
	if h.today.Operators == nil {
		return roster{}, nil
	}
	all, err := h.today.Operators(ctx)
	if err != nil {
		return roster{}, err
	}

	var out roster
	for _, one := range all {
		if one.RevokedAt != nil {
			continue
		}
		if !one.SecondFactor {
			out.noFactor++
		}
		if one.LastOpenedConsole == nil {
			out.neverUsed++
		}
	}
	return out, nil
}

type nights struct{ failed, adrift, never int }

/*
jobs counts what the scheduled work did last, one job at a time.

	THE LATEST RUN AND NOT THE HISTORY. "Did it work last night" is a question
	about the most recent attempt: a job that failed on Tuesday and has run
	every night since is a job that is fine, and a screen that counted failures
	would keep reporting Tuesday for ever.

	A JOB THAT HAS RECORDED NOTHING IS A FINDING AND NOT A ZERO. `Names` reads
	what has ever recorded a run, so a job that never has does not appear there
	at all — which is why the startable list is asked as well. A scheduled job
	with no runs is either newly deployed or a scheduler that never fired, and
	both want a person.
*/
func (h *TodayHandler) jobs(ctx context.Context) (nights, error) {
	var out nights
	if h.today.Jobs.Names == nil || h.today.Jobs.Latest == nil {
		return out, nil
	}

	ran, err := h.today.Jobs.Names(ctx)
	if err != nil {
		return out, err
	}

	seen := map[string]bool{}
	for _, name := range ran {
		seen[name] = true

		latest, err := h.today.Jobs.Latest(ctx, name, 1)
		if err != nil {
			return out, fmt.Errorf("reading the last run of %s: %w", name, err)
		}
		if len(latest) == 0 {
			continue
		}
		switch one := latest[0]; {
		case one.Adrift:
			out.adrift++
		case one.Outcome == "failed":
			out.failed++
		}
	}

	for _, name := range h.today.Jobs.Startable {
		if !seen[name] {
			out.never++
		}
	}
	return out, nil
}
